crates/wasm-smith/src/core.rs

```diff
@@ -938,8 +938,23 @@ impl Module {
     ) -> Result<HeapType> {
         use {AbstractHeapType as AHT, CompositeInnerType as CT, HeapType as HT};
 
+        // If GC is disabled then favor the "top" type of the input type.
         if !self.config.gc_enabled {
-            return Ok(ty);
+            return Ok(match ty {
+                HeapType::Concrete(i) => {
+                    let CompositeType { inner, shared } = &self.ty(i).composite_type;
+                    let ty = match inner {
+                        CompositeInnerType::Array(_) => AbstractHeapType::Array,
+                        CompositeInnerType::Struct(_) => AbstractHeapType::Struct,
+                        CompositeInnerType::Func(_) => AbstractHeapType::Func,
```

Comment:
The only case that this would be valid for is `(ref null? func)`. If GC is disabled, then struct and array references will still be invalid.

Which again points me towards this trying to do something that doesn't make sense...